RFC-0022-Model-Partitioning-in-Pipeline-Parallelism.md

```diff
@@ -0,0 +1,255 @@
+
+
+```
+    =========  =============  =========================================
+    Optimizer  `param_scale`  Internal State
+    =========  =============  =========================================
+    SGD        2--3           (momentum_buffer)
+    Adam       4--5           exp_avg, exp_avg_sq, (max_exp_avg_sq)
+    Adadelta   4              square_avg, acc_delta
+    Adagrad    3              sum
+    RMSprop    3--5           square_avg, (momentum_buffer), (grad_avg)
+    =========  =============  =========================================
+```
+
+
+Since we assigned the cost to each top-level module, we can do the partition, basic partition function will be like the below:
+
+```
+def _partition_by_cost(interp: torch.fx.Interpreter,
+                       num_partitions):
+    submodule_costs = interp.get_costs() 
+    # Dict[module_qual_name, cost], execution order
+    
+    # use Block Partition of Sequences to minimize the variance
+    # of the submodules' cost list 
+    partitions = 
+        block_partition.solve(submodule_costs, num_partitions)
+    return partitions
+```
+
+
+Q: How do we collect the communication costs?
+
+* activation_size and gradient size, could potentially contribute to part of the communication cost
+* What about the device info? how do we know each pairs are NVLink or PCIe?
+
+
+
+### Further exploration of partition algorithm
+
+Since we hide the underlying partition algorithm, we can do further partition algorithm explorations, some potential explorations we can do:
+
+* exploration like in DAPPLE when using non-simple schedule (i.e. we reduce the bubble by do the backward as early as possible). 
+* Using torch.fx, we can trace the module architecture in module level, and possibly split the entire traced graph into several submodules (this might not fully resemble the existing module architecture). 
+* DP-based algorithm like SageMaker, but could be on operation-level with more granularity 
+
+
+
+## Potential issues
+
+**What if a model with a big memory requirement that couldn’t fit into a single GPU?**
+A: we construct the model on CPU and do fx tracing on CPU, after the partition we move it to the corresponding device
+
+**What if a “module” (submodule) with a big memory requirement that couldn’t fit into a single GPU?**
+Can we use ShardedTensor to shard the module parameters? or we could do operation level tracing, partition this module into 2 submodules, then assign to different devices. 
+
+**What if constructing the model on CPU itself is hard?** 
+User created the model first on meta device, we use the model that’s not materializing to do symbolic tracing, but we couldn’t do profiling since it’s not materialized yet, after we do symbolic tracing, we should use a simple partition algorithm (i.e. only count the param sizes) to do the partition, then materialize afterwards.
```

Comment:
Is there a documentation on "meta" device? I don't seem to find any.

Reply:
https://github.com/pytorch/rfcs/blob/rfc-0005/RFC-0005-structured-kernel-definitions.md